### PR TITLE
Allowed Netty Epoll/Kqueue to work in IPv6 Only environments

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Motivation:
In some environments, IPv4 may be disabled (at a kernel level).
Google has such an environment for testing v4 -> v6 transition
paths.  This give confidence that code is v6 ready.

Modifications:
Change native socket code to ignore failures of trying to enter
dual stack mode.  This change has been made to Google's internal
JDK, and will/should be upstreamed to OpenJDK eventually.

Results:
Netty works in IPv6 only environments

Fixes: #6993

### CODE REVIEW

File: transport-native-unix-common/src/main/c/netty_unix_socket.c [hunk 1]
```diff
@@ -188,17 +188,48 @@ static int socket_type(JNIEnv* env) {
...
         int optval = 0;
-        if (netty_unix_socket_setOption(env, fd, IPPROTO_IPV6, IPV6_V6ONLY, &optval, sizeof(optval)) < 0) {
-            // Something went wrong so close the fd and return here. setOption(...) itself throws the exception already.
-            close(fd);
-            return -1;
+        if (netty_unix_socket_setOption0(fd, IPPROTO_IPV6, IPV6_V6ONLY, &optval, sizeof(optval)) < 0) {
+            if (errno == EAFNOSUPPORT) {
+              // Something went wrong, but don't close fd and don't return -1.   At best we can do is log.
+              // TODO: bubble this up to an actual Logger.
+              fprintf(stderr, "setsockopt() failed: Unable to set IPV6_V6ONLY.  Continuing.\n");
```
Comment (normanmaurer) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Please address the TODO

Comment (normanmaurer) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Or just remove the fprintf for now

File: transport-native-unix-common/src/main/c/netty_unix_socket.c [hunk 2]
```diff
@@ -289,26 +320,10 @@ static jobject _recvFrom(JNIEnv* env, jint fd, void* buffer, jint pos, jint limi
     return createDatagramSocketAddress(env, &addr, res);
 }
 
-static void netty_unix_socket_optionHandleError(JNIEnv* env, int err, char* method) {
-    if (err == EBADF) {
```
Comment (normanmaurer) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Please revert these changes as it seems this just moves code and so is not related

Comment (Scottmitch) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: IIUC the static methods have to move up to avoid their first use case or we need to declare the method before the first use.

File: transport-native-unix-common/src/main/c/netty_unix_socket.c [hunk 3]
```diff
@@ -321,14 +336,6 @@ int netty_unix_socket_getOption(JNIEnv* env, jint fd, int level, int optname, vo
     return rc;
 }
 
-int netty_unix_socket_setOption(JNIEnv* env, jint fd, int level, int optname, const void* optval, socklen_t len) {
```
Comment (Scottmitch) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: IIUC this method doesn't have to move (it isn't static and is declared in the header file)?

File: transport-native-unix-common/src/main/c/netty_unix_socket.c (hunk 1 above)
Comment (Scottmitch) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Why do we want to suppress the error here? `socket_type` accounts for `EAFNOSUPPORT` so if we get it here it seems like a programming error. Can we just propagate the error via `netty_unix_socket_setOptionHandleError` in this case too?

File: transport-native-unix-common/src/main/c/netty_unix_socket.c (hunk 2 above)
Comment (carl-mastrangelo) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: netty_unix_socket_setOption0 needs to be defined before _socket() uses it, and I wanted to keep most of these functions together.  

I can move the minimal number of functions to do so, or move _socket down.  Preference?

File: transport-native-unix-common/src/main/c/netty_unix_socket.c (hunk 1 above)
Comment (carl-mastrangelo) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: done.

File: transport-native-unix-common/src/main/c/netty_unix_socket.c (hunk 2 above)
Comment (Scottmitch) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Moving just the static methods sgtm

File: transport-native-unix-common/src/main/c/netty_unix_socket.c [hunk 4]
```diff
@@ -188,17 +188,47 @@ static int socket_type(JNIEnv* env) {
...
     } else if (socketType == AF_INET6) {
-        // Allow to listen /connect ipv4 and ipv6
+        // Try to allow listen /connect ipv4 and ipv6
         int optval = 0;
-        if (netty_unix_socket_setOption(env, fd, IPPROTO_IPV6, IPV6_V6ONLY, &optval, sizeof(optval)) < 0) {
-            // Something went wrong so close the fd and return here. setOption(...) itself throws the exception already.
-            close(fd);
-            return -1;
+        if (netty_unix_socket_setOption0(fd, IPPROTO_IPV6, IPV6_V6ONLY, &optval, sizeof(optval)) < 0) {
+            if (errno == EAFNOSUPPORT) {
```
Comment (Scottmitch) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Why do we want to suppress the error here? socket_type accounts for EAFNOSUPPORT so if we get it here it seems like a programming error. Can we just propagate the error via netty_unix_socket_setOptionHandleError in this case too?

Comment (Scottmitch) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: This was collapsed in https://github.com/netty/netty/pull/6994#discussion_r128362622

File: transport-native-unix-common/src/main/c/netty_unix_socket.c (hunk 3 above)
Comment (carl-mastrangelo) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Sure, it doesn't have to, but some teams prefer to keep related code adjacent / contiguous if possible.  If you don't have such a requirement I can move the bare minimum.

File: transport-native-unix-common/src/main/c/netty_unix_socket.c (hunk 4 above)
Comment (carl-mastrangelo) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: That code checks if AF_INET6 is allowed, but not single stack mode which is what `_socket()` is asking for.

File: transport-native-unix-common/src/main/c/netty_unix_socket.c (hunk 2 above)
Comment (carl-mastrangelo) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Done.

File: transport-native-unix-common/src/main/c/netty_unix_socket.c (hunk 4 above)
Comment (Scottmitch) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Can we remote the `if` since it doesn't do anything? What about the following:

```c
if (errno != EAFNOSUPPORT) {
...
}
// else we failed to enable dual stack mode.
// It is assumed the socket is re‐stricted to sending and receiving IPv6 packets only.
```

Comment (carl-mastrangelo) [inline transport-native-unix-common/src/main/c/netty_unix_socket.c]: Done.

**Comment by Scottmitch** [comment]
4.1 (3d22b2424)

I will not back port this to 4.0 because the code has diverged in this area.

**Comment by normanmaurer** [comment]
@Scottmitch not backporting sounds good.
